### PR TITLE
feat/improve test setup

### PR DESCRIPTION
**Pull request by brianmcgee**
## Overview

- NixOS module tests can now be co-located with their respective modules. 
- Added mission control script for running specific tests instead of needing to use `nix flake check`
- Packages are now prefixed with `package-` and tests are prefixed with `test-` within the checks attribute set.
- Updated the development section of the readme

## Testing

Any file within the `modules` directory tree with a suffix of `.test.nix` is automatically detected and added to `config.testing.checks` which is an attribute set of packages for mixing in with the main flake checks. 

The test file must be of the form:

```nix
{
	# used for filtering out when running on unsupported platforms
	systems = [ "x86_64-linux" ];

	# the main test module
	module = { ... }: {

    };
}
```

You can see an example in `modules/clients/execution/nethermind/basic.test.nix`.

### CODE REVIEW

File: modules/clients/execution/nethermind/basic.test.nix [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+
+    nodes = {
+      basicConf = {
+        self,
+        system,
+        lib,
+        ...
+      }: {
+        imports = [
+          self.nixosModules.default
```
Comment (brianmcgee) [inline modules/clients/execution/nethermind/basic.test.nix:17]: Unfortunately this is necessary. I tried importing this `modules/testing.nix` before importing the test module but it wouldn't work for some reason.

File: modules/clients/execution/nethermind/basic.test.nix [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+        ];
+
+        # see: https://docs.nethermind.io/nethermind/first-steps-with-nethermind/system-requirements
+        virtualisation.cores = 2;
+        virtualisation.memorySize = 8192;
+
+        services.ethereum.nethermind.sepolia = {
+          enable = true;
+          package = self.packages.${system}.nethermind;
```
Comment (brianmcgee) [inline modules/clients/execution/nethermind/basic.test.nix:26]: Unfortunately this is necessary. I tried constructing a custom nixpkgs with our overlay before loading the test module in `modules/testing.nix` but it wasn't detecting our flake packages. This is a workaround in the meantime.

File: modules/clients/execution/nethermind/basic.test.nix (hunk 1 above)
Comment (brianmcgee) [reply modules/clients/execution/nethermind/basic.test.nix]: Fixed this by setting `defaults.imports = [ ]` in `runTest`.

File: modules/clients/execution/nethermind/basic.test.nix (hunk 2 above)
Comment (brianmcgee) [reply modules/clients/execution/nethermind/basic.test.nix]: Fixed this by passing custom pkgs with overlay into `node.specialArgs` in `runTest`

File: modules/testing.nix [hunk 3]
```diff
@@ -0,0 +1,145 @@
+{
+  self,
+  inputs,
+  lib,
```
Comment (aldoborrero) [inline modules/testing.nix:4]: @brianmcgee is this lib attr necessary?

File: modules/testing.nix [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+    ## Commands
+    ########################################
+    config.mission-control.scripts = let
+      category = "Testing";
+    in {
+      test = {
+        inherit category;
+        description = "Build and run a test";
+        exec = ''
+
```
Comment (aldoborrero) [inline modules/testing.nix:90]: @brianmcgee wouldn't make sense to set `set -eo pipefail` at the beginning of this script?

**Review by aldoborrero — DISMISSED** [review]
@brianmcgee minor comments for you to review. Otherwise LGTM.

File: modules/testing.nix (hunk 3 above)
Comment (brianmcgee) [reply modules/testing.nix]: Removed

File: modules/testing.nix (hunk 4 above)
Comment (brianmcgee) [reply modules/testing.nix:90]: Added

**Review by aldoborrero — APPROVED** [review]

**Review by selfuryon — APPROVED** [review]
LGTM for me, really cool changes for testing

**Pull request merged**